master/internal/resourcemanagers/agent/agent_state.go

```diff
@@ -378,3 +414,274 @@ func (a *AgentState) patchSlotState(
+		if rpConfig.AgentReattachEnabled {
+			result = append(result, rpConfig.PoolName)
+		}
+	}
+
+	return result
+}
+
+// retrieveAgentStates reconstructs AgentStates from the database for all resource pools that
+// have agent_container_reattachment enabled.
+func retrieveAgentStates() (map[AgentID]AgentState, error) {
+	rpNames := listResourcePoolsWithReattachEnabled()
+
+	if len(rpNames) == 0 {
+		return map[AgentID]AgentState{}, nil
+	}
+
+	snapshots := []AgentSnapshot{}
+	err := db.Bun().NewSelect().Model(&snapshots).
+		Where("resource_pool_name IN (?)", bun.In(rpNames)).
+		Scan(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[AgentID]AgentState, len(snapshots))
+
+	for _, s := range snapshots {
+		state, err := newAgentStateFromSnapshot(s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to recreate agent state %s: %w", s.AgentID, err)
+		}
+
+		result[s.AgentID] = *state
+	}
+
+	return result, nil
+}
+
+func newAgentStateFromSnapshot(as AgentSnapshot) (*AgentState, error) {
+	parsedUUID, err := uuid.Parse(as.UUID)
+	if err != nil {
+		return nil, err
+	}
+
+	slotStates := make(map[device.ID]*slot)
+	devices := make(map[device.Device]*cproto.ID)
+
+	for _, sd := range as.Slots {
+		slotStates[sd.Device.ID] = &slot{
+			device:      sd.Device,
+			containerID: sd.ContainerID,
+			enabled: slotEnabled{
+				deviceAdded:  true,
+				agentEnabled: as.UserEnabled,  // TODO
+				userEnabled:  as.UserEnabled,  // TODO
+				draining:     as.UserDraining, // TODO
+			},
+		}
+		if sd.ContainerID != nil {
+			devices[sd.Device] = sd.ContainerID
+		} else {
+			devices[sd.Device] = nil
+		}
+	}
+
+	containerState := make(map[cproto.ID]*cproto.Container)
+
+	if len(as.Containers) > 0 {
+		containerSnapshots := make([]ContainerSnapshot, 0, len(as.Containers))
+		err := db.Bun().NewSelect().Model(&containerSnapshots).
+			Where("container_id IN (?)", bun.In(as.Containers)).
+			Scan(context.TODO())
+		if err != nil {
+			return nil, err
+		}
+
+		for _, containerSnapshot := range containerSnapshots {
+			container := containerSnapshot.ToContainer()
+			containerState[container.ID] = &container
+		}
+	}
+
+	result := AgentState{
+		maxZeroSlotContainers: as.MaxZeroSlotContainers,
+		resourcePoolName:      as.ResourcePoolName,
+		Label:                 as.Label,
+		uuid:                  parsedUUID,
+		enabled:               as.UserEnabled,
+		draining:              as.UserDraining,
+		slotStates:            slotStates,
+		Devices:               devices,
+		containerAllocation:   make(map[cproto.ID]*actor.Ref),
+		containerState:        containerState,
+	}
+
+	return &result, nil
+}
+
+func (a *AgentState) restoreContainersField() error {
+	containerIDs := maps.Keys(a.containerState)
+
+	c2a, err := loadContainersToAllocationIds(containerIDs)
+	if err != nil {
+		return err
+	}
+
+	containers := make(map[cproto.ID]*actor.Ref)
+	for contID := range c2a {
+		ref := task.GetAllocation(c2a[contID])
```

Comment:
more just curious, why not
```suggestion
	for contID, alloc := range c2a {
		ref := task.GetAllocation(alloc)
```